skopt/optimizer/optimizer.py

```diff
@@ -0,0 +1,242 @@
+        is updated with the optimal value obtained by optimizing `acq_func`
+        with `acq_optimizer`.
+
+        - If set to `"sampling"`, then `acq_func` is optimized by computing
+          `acq_func` at `n_points` sampled randomly.
+        - If set to `"lbfgs"`, then `acq_func` is optimized by
+              - Sampling `n_restarts_optimizer` points randomly.
+              - `"lbfgs"` is run for 20 iterations with these points as initial
+                points to find local minima.
+              - The optimal of these local minima is used to update the prior.
+        - If set to `"auto"`, then it is set to `"lbfgs"`` if
+          all the search dimensions are Real (continuous). It defaults to
+          `"sampling"` for all other cases.
+
+    * `random_state` [int, RandomState instance, or None (default)]:
+        Set random state to something other than None for reproducible
+        results.
+
+    * `n_points` [int, default=500]:
+        Number of points to sample to determine the next "best" point.
+        Useless if acq_optimizer is set to `"lbfgs"`.
+
+    * `n_restarts_optimizer` [int, default=5]:
+        The number of restarts of the optimizer when `acq_optimizer`
+        is `"lbfgs"`.
+
+    * `xi` [float, default=0.01]:
+        Controls how much improvement one wants over the previous best
+        values. Used when the acquisition is either `"EI"` or `"PI"`.
+
+    * `kappa` [float, default=1.96]:
+        Controls how much of the variance in the predicted values should be
+        taken into account. If set to be very high, then we are favouring
+        exploration over exploitation and vice versa.
+        Used when the acquisition is `"LCB"`.
+
+    * `n_jobs` [int, default=1]
+        Number of cores to run in parallel while running the lbfgs
+        optimizations over the acquisition function. Valid only when
+        `acq_optimizer` is set to "lbfgs."
+        Defaults to 1 core. If `n_jobs=-1`, then number of jobs is set
+        to number of cores.
+    """
+    def __init__(self, dimensions, base_estimator,
+                 n_random_starts=10, acq_func="EI", acq_optimizer="auto",
+                 random_state=None, n_points=10000, n_restarts_optimizer=5,
+                 xi=0.01, kappa=1.96, n_jobs=1):
+        super(Optimizer, self).__init__()
+        # Arguments that are just stored not checked
+        self.acq_func = acq_func
+        self.rng = check_random_state(random_state)
+        self.kappa = kappa
+        self.models = []
+        self.n_points = n_points
+        self.n_restarts_optimizer = n_restarts_optimizer
+        self.space = Space(dimensions)
+        self.Xi = []
+        self.xi = xi
+        self.yi = []
+
+        self._check_arguments(dimensions, base_estimator,
+                              n_random_starts,
+                              acq_func, acq_optimizer,
+                              n_jobs)
+
+        self.parallel = Parallel(n_jobs=n_jobs)
```

Comment:
I would prefer keep things separated as Tim proposed in this PR. Evaluating the function within this class `Optimizer` would make the ask and tell interface inconsistent in my opinion (i.e., if the `Optimizer` object had access to the objective function, then there would be no point in having a `tell` method at all.)